lib/collection/src/wal_delta.rs

```diff
@@ -279,6 +280,119 @@ mod tests {
+            // Create operation on peer A
+            let mut a_clock_0 = a_clock_set.get_clock();
+            let clock_tag = ClockTag::new(1, a_clock_0.id(), a_clock_0.tick_once());
+            let operation = OperationWithClockTag::new(mock_operation(op_id), Some(clock_tag));
+
+            // Write operation to peers A, B and C
+            for wal in [&a_wal, &b_wal, &c_wal] {
+                let mut operation = operation.clone();
+                let (_, _) = wal.lock_and_write(&mut operation).await.unwrap();
+                a_clock_0.advance_to(operation.clock_tag.unwrap().clock_tick);
+            }
+        }
+
+        // Now, let's assume that node B is started a stream records transfer to node C,
+        // and so there's a forward proxy on the node B that forwards all operations to the node C.
+        //
+        // (Because we are not using real shards in this test, we will "simulate" the forward proxy
+        // by sending each operation to the node C twice.)
+        //
+        // And, let's assume that we restarted the node A, and so clock set on node A is reset to
+        // all `0`s.
+        let mut a_clock_set = ClockSet::new();
+        let mut a_clock_0 = a_clock_set.get_clock();
+
+        // Let's create another operation on peer A and write it on all nodes:
+
+        // Create operation on peer A
+        let clock_tag = ClockTag::new(1, a_clock_0.id(), a_clock_0.tick_once());
+        let operation = OperationWithClockTag::new(mock_operation(4), Some(clock_tag));
+
+        // Write operation on peers A and B
+        for wal in [&a_wal, &b_wal] {
+            let mut operation = operation.clone();
+            let (_, _) = wal.lock_and_write(&mut operation).await.unwrap();
+            a_clock_0.advance_to(operation.clock_tag.unwrap().clock_tick);
+        }
+
+        // Then write operation on peer C from the forward proxy *first*
+        {
+            let mut operation = operation.clone();
+            let (_, _) = c_wal.lock_and_write(&mut operation).await.unwrap();
+            a_clock_0.advance_to(operation.clock_tag.unwrap().clock_tick);
+        }
+
+        // And now let's assume that operation, that was sent from node A to C directly was written
+        // to WAL, but the response, that should have advanced the clock set on node A failed.
+        //
+        // Currently, we simply ignore all errors from the nodes in `Partial` state, so it won't
+        // mark node C as `Dead`.
+        {
+            let mut operation = operation.clone();
+            let (_, _) = c_wal.lock_and_write(&mut operation).await.unwrap();
+            // we do not advance `a_clock_0` here
+        }
+
+        // Now, if we assume that:
+        // - we increment local clock map, when we receive an operation with `clock_tick = 0`
+        // - and we reject older operations
+        //   - (and we have to do so, otherwise we can't guarantee that our WAL has monotoncally increasing versions)
```

Comment:
> has monotoncally increasing versions

This is not the property we test. It can actually go down. We only want to make sure that the end of our WAL has the highest clocks.

Reply:
That was what I meant, it was just annoying to formulate properly.